ls1intum/Artemis
# `General`: Improve performance when deleting exercises

### Checklist
#### General


#### Server

### Motivation and Context



When deleting or resetting an exercise with potentially thousands of results, we were scheduling an update of the corresponding participant score for each of those.
This was unneeded, as when the exercise is deleted, we can also remove all participant scores related to it.
Additionally, there was a bug where the scheduler tried to update a participant score while all results for the participation were in the process of being deleted (resulting in a deadlock).

When testing this, we found out that deleting an exercise generally (even before the changes to the participant scores) takes up to five minutes! ⏳ 

### Description


The performance was improved by simply removing the participation from the result before it is deleted. The listener will ignore the result in these cases and _not_ schedule a task to update the participant score.
Furthermore, we now check if the participant score is already deleted when a task is scheduled because a result is (about to be) removed. This solves the deadlock problem.

Last but not least, Stephan managed to remove all `@Transactional` annotations from the exercise deletion, which speeds up the process by a factor of ten.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Exercise with many participations and (thousands of) results

1. Log in to Artemis
2. Navigate to Course Administration
3. Delete or reset the exercise
4. Ensure no exceptions occur in the log
5. Also make sure the existing functionality (updating the participant score when a result is created/update) still works

### Review Progress





#### Code Review
#### Manual Tests
